builder/builder.go

```diff
@@ -0,0 +1,457 @@
+func (b *Builder) listenPool(ctx context.Context) error {
+	for {
+		if err := b.depletePool(ctx); err != nil {
+			if !errors.Is(err, mempool.ErrTxnPoolEmpty) {
+				return err
+			}
+		}
+
+		// push the pending block to the feed
+		b.subPendingBlock.Send(b.PendingBlock())
+		select {
+		case <-ctx.Done():
+			return nil
+		// We wait for the mempool to get more txns before we continue
+		case <-b.mempool.Wait():
+			continue
+		}
+	}
+}
+
+// depletePool pops all available transactions from the mempool,
+// and executes them in sequence, applying the state changes
+// to the pending state
+func (b *Builder) depletePool(ctx context.Context) error {
+	blockHashToBeRevealed, err := b.getRevealedBlockHash()
+	if err != nil {
+		return err
+	}
+	for {
+		b.finaliseMutex.RLock()
+		userTxn, err := b.mempool.Pop()
+		if err != nil {
+			b.finaliseMutex.RUnlock()
+			return err
+		}
+		b.log.Debugw("running txn", "hash", userTxn.Transaction.Hash().String())
+		if err = b.runTxn(&userTxn, blockHashToBeRevealed); err != nil {
+			b.log.Debugw("failed txn", "hash", userTxn.Transaction.Hash().String(), "err", err.Error())
+			var txnExecutionError vm.TransactionExecutionError
+			if !errors.As(err, &txnExecutionError) {
+				b.finaliseMutex.RUnlock()
+				return err
+			}
+		}
+		b.log.Debugw("running txn success", "hash", userTxn.Transaction.Hash().String())
+		b.finaliseMutex.RUnlock()
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+	}
+}
+
+func (b *Builder) getRevealedBlockHash() (*felt.Felt, error) {
+	blockHeight, err := b.bc.Height()
+	if err != nil {
+		return nil, err
+	}
+	const blockHashLag = 10
+	if blockHeight < blockHashLag {
+		return nil, nil
+	}
+
+	header, err := b.bc.BlockHeaderByNumber(blockHeight - blockHashLag)
+	if err != nil {
+		return nil, err
+	}
+	return header.Hash, nil
+}
+
+// runTxn executes the provided transaction and applies the state changes
+// to the pending state
+func (b *Builder) runTxn(txn *mempool.BroadcastedTransaction, blockHashToBeRevealed *felt.Felt) error {
+	// Get the pending state
+	pending, err := b.Pending()
+	if err != nil {
+		return err
+	}
+
+	// Create a state writer for the transaction execution
+	state := sync.NewPendingStateWriter(pending.StateUpdate.StateDiff, pending.NewClasses, b.headState)
+
+	// Prepare declared classes, if any
+	var declaredClass []core.Class
+	if txn.DeclaredClass != nil {
+		declaredClass = append(declaredClass, txn.DeclaredClass)
+	}
+
+	paidFeesOnL1 := []*felt.Felt{}
+	if txn.PaidFeeOnL1 != nil {
+		paidFeesOnL1 = append(paidFeesOnL1, txn.PaidFeeOnL1)
+	}
+	// Execute the transaction
+	vmResults, err := b.vm.Execute(
+		[]core.Transaction{txn.Transaction},
+		declaredClass,
+		paidFeesOnL1,
+		&vm.BlockInfo{
+			Header:                pending.Block.Header,
+			BlockHashToBeRevealed: blockHashToBeRevealed,
+		},
+		state,
+		b.bc.Network(),
+		false, false, false, true)
+	if err != nil {
+		return err
+	}
```

Comment:
The performance gains come here where everything in this function before and after this block of code gets executed only once instead of `n` where `n` is the amount of transactions in the pool at the moment